airflow/kubernetes/pod_launcher.py

```diff
@@ -172,6 +181,50 @@ def read_pod(self, pod):
                 'There was an error reading the kubernetes API: {}'.format(e)
             )
 
+    def _handle_istio_proxy(self, pod):
+        """If an istio-proxy sidecar is detected, attempt to cleanly shutdown.
+        If we detect a version of Istio before it's compatible with Kubernetes
+        Jobs, then raise an informative error message.
+
+        Args:
+            pod (V1Pod): The pod which we are checking for the sidecar
+
+        Returns:
+            (bool): True if we detect and exit istio-proxy,
+                    False if we do not detect istio-proxy
+
+        Raises:
+            AirflowException: if we find an istio-proxy, and we can't shut it down.
+        """
+        # describe the pod
+        pod = self.read_pod(pod)
+        for container in pod.spec.containers:
+            if container.name != SidecarNames.ISTIO_PROXY:
+                continue
+            # Check if supported version of istio-proxy.
+            # If we can't tell the version, proceed anyways.
+            if ":" in container.image:
+                _, tag = container.image.split(":")
+                if semantic_version(tag) < semantic_version("1.3.0-rc.0"):
+                    raise AirflowException(
+                        'Please use istio version 1.3.0+ for KubeExecutor compatibility.' +\
+                        ' Detected version {}'.format(tag))
+            #  exec into the container
+            resp = kubernetes_stream(self._client.connect_get_namespaced_pod_exec,
```

Comment:
Is there no way to make the call to the sidecar directly from the airflow pod, without `exec`ing into the sidecar?

Reply:
I wonder if you can stop using streams here. Websocket and writing to the stream adds extra complexity. We can run the command using the "cmd" parameter eg. `["/bin/sh", "-c", "curl http://127.0.0.1:15020/quitquitquit"]`

Reply:
Referencing directly to Pod will be very problematic because of network isolation. First we would have to redirect ports. I think the use of exec is fine ..

Reply:
Is 15020 the default port for the istio proxy? Where I would be a little concerned Is whether this would change? Otherwise I would just want this well documented since it's not as readily apparent as reading from the kube API

Reply:
@mik-laj what are you referring to when you say "extra complexity"? Does it make k8s slower? Does it do a lot more networking?

Reply:
Creating a websocket connection requires sending one HTTP request, and then a statefull connection is maintained. Ithink, It is not necessary. It seems to me that all we have to do is send a request to execute the command.  I suspect that in this case it will not slow down, but maintaining a large number of statefull connections can cause a problem. This definitely increases traffic, but very little.

Reply:
I will do experiments and check it.

Reply:
> Is there no way to make the call to the sidecar directly from the airflow pod, without `exec`ing into the sidecar?

> I wonder if you can stop using streams here. Websocket and writing to the stream adds extra complexity. We can run the command using the "cmd" parameter eg. `["/bin/sh", "-c", "curl http://127.0.0.1:15020/quitquitquit"]`

I agree the direct exec is less complicated in this case instead of exec in, then write to stream. Let me give that a shot, thank you.

> Is 15020 the default port for the istio proxy? Where I would be a little concerned Is whether this would change? Otherwise I would just want this well documented since it's not as readily apparent as reading from the kube API

Yeah, it's the default port. Allow me to consider how I might grab it programmatically, for example it might be in the pod spec. Thanks! Regarding the extra complexity, I think @mik-laj just means that it's two steps when it could be one, correct me if I'm wrong!

Reply:
When we do not use the stream, we immediately call the command instead of typing on stdin.

Reply:
15020 is the status port for istio, and that's where they have chosen to put the /quitquitquit route. https://github.com/istio/istio/pull/15406/files I can find it as a command line option to istio-proxy in my pod spec:
```
  istio-proxy:
    Image:         istio/proxyv2:1.3.0-rc.1
...
    Port:          15090/TCP # for clarity, it is not the service port on the pod. it is internal only
    Args:
      proxy
      sidecar
 ...
      --statusPort
      15020
...
```
and in the readiness probe:
```
Readiness:  http-get http://:15020/healthz/ready delay=1s timeout=1s period=2s #success=1 #failure=30
```
it's also available in the helm values, should helm have been used.
```
https://github.com/istio/istio/blob/3a955658f9a10153ec1e056d6e1e03f1870e4dcc/install/kubernetes/helm/istio/values.yaml
```
I think the best and most authoritative option here is the command line argument. I could attempt to parse it from the pod spec and then fall back to default in the even of a parse failure / can't find it.